File: src/monitor.rs
use std::time::Duration;

use crate::error::Result;
use crate::kill;
use crate::memory;
use crate::memory::MemoryInfo;
use crate::process::Process;

enum MemoryStatus {
    NearTerminal(f32),
    Okay
}

pub struct Monitor {
    memory_info: MemoryInfo,
    proc_buf: [u8; 50],
    buf: [u8; 100],
    status: MemoryStatus
}

impl Monitor {
    /// Determines how much oomf should sleep
    /// This function is essentially a copy of how earlyoom calculates its sleep time
    ///
    /// Credits: https://github.com/rfjakob/earlyoom/blob/dea92ae67997fcb1a0664489c13d49d09d472d40/main.c#L365
    /// MIT Licensed
    pub fn sleep_time_ms(&self) -> Duration {
        // Maximum expected memory fill rate as seen
        // with `stress -m 4 --vm-bytes 4G`
        const RAM_FILL_RATE: i64 = 6000;
        // Maximum expected swap fill rate as seen
        // with membomb on zRAM
        const SWAP_FILL_RATE: i64 = 800;

        // Maximum and minimum time to sleep, in ms.
        const MIN_SLEEP: i64 = 100;
        const MAX_SLEEP: i64 = 1000;

        // TODO: make these percentages configurable by args./config. file
        const RAM_TERMINAL_PERCENT: f64 = 10.;
        const SWAP_TERMINAL_PERCENT: f64 = 10.;

        let ram_headroom_kib = (self.memory_info.available_ram_percent as f64
            - RAM_TERMINAL_PERCENT)
            * (self.memory_info.total_ram_mb as f64 * 10.0);
        let swap_headroom_kib = (self.memory_info.available_swap_percent as f64
            - SWAP_TERMINAL_PERCENT)
            * (self.memory_info.total_swap_mb as f64 * 10.0);

        let ram_headroom_kib = i64::max(ram_headroom_kib as i64, 0);
        let swap_headroom_kib = i64::max(swap_headroom_kib as i64, 0);

        let time_to_sleep = ram_headroom_kib / RAM_FILL_RATE + swap_headroom_kib / SWAP_FILL_RATE;
        // let time_to_sleep = i64::min(time_to_sleep, MAX_SLEEP);
        let time_to_sleep = i64::max(time_to_sleep, MIN_SLEEP);

        Duration::from_millis(time_to_sleep as u64)
    }

    pub fn new(proc_buf: [u8; 50], mut buf: [u8; 100]) -> Result<Self> {
        let memory_info = MemoryInfo::new()?;
        let status = if memory_info.available_ram_percent <= 15 {
            MemoryStatus::NearTerminal(memory::pressure::pressure_some_avg10(&mut buf)?)
        } else {
            MemoryStatus::Okay
        };

        Ok(Self { memory_info, proc_buf, buf, status })
    }

    fn memory_is_low(&self) -> bool {
        // TODO: doing this just for testing.
        // Must account for swap later on and
        // allow the terminal percentage to be
        // modified by command-line arg. and/or config. file

        matches!(self.status, MemoryStatus::NearTerminal(psi) if psi >= 30.0)
    }

    fn get_victim(&mut self) -> Result<Process> {
        kill::choose_victim(&mut self.proc_buf, &mut self.buf)
    }

    fn update_memory_stats(&mut self) -> Result<()> {
        self.memory_info = memory::MemoryInfo::new()?;
        self.status = if self.memory_info.available_ram_percent <= 15 {
            let psi = memory::pressure::pressure_some_avg10(&mut self.buf)?;
            eprintln!("Near terminal! PSI: {}", psi);
            MemoryStatus::NearTerminal(psi)
        } else {
            MemoryStatus::Okay
        };
        Ok(())
    }

    fn free_up_memory(&mut self) -> Result<()> {
        let victim = self.get_victim()?;
        
        // TODO: is this necessary?
        //
        // Check for memory stats again to see if the
        // low-memory situation was solved while
        // we were searching for our victim
        self.update_memory_stats()?;
        if self.memory_is_low() {
            kill::kill_and_wait(victim)?;
        }
        Ok(())
    }

    // Use the never type here whenever it reaches stable
    pub fn poll(&mut self) -> Result<()> {
        loop {
            // Update our memory readings
            self.update_memory_stats()?;
            if self.memory_is_low() {
                self.free_up_memory()?;
            }
            // Adaptive sleep time
            let sleep_time = self.sleep_time_ms();
            // eprintln!("Sleeping {}ms", sleep_time.as_millis());
            
            std::thread::sleep(sleep_time);
        }
        Ok(())
    }
}

File: zig/src/monitor.zig
const std = @import("std");
const time = std.time;
const math = std.math;

const memory = @import("memory.zig");
const pressure = @import("pressure.zig");
const process = @import("process.zig");
const config = @import("config.zig");

const MemoryStatusTag = enum {
    ok,
    near_terminal,
};

const MemoryStatus = union(MemoryStatusTag) {
    /// Memory is "okay": basically no risk of memory thrashing
    ok: void,
    /// Nearing the terminal PSI cutoff: memory thrashing is occurring or close to it. Holds the current PSI value.
    near_terminal: f32,
};

pub const Monitor = struct {
    mem_info: memory.MemoryInfo,
    /// Memory status as of last checked
    status: MemoryStatus,
    /// A pointer to a buffer of at least 128 bytes
    buffer: []u8,
    const Self = @This();

    pub fn new(buffer: []u8) !Self {
        var self = Self{
            .mem_info = undefined,
            .status = undefined,
            .buffer = buffer,
        };

        try self.updateMemoryStats();

        return self;
    }

    pub fn updateMemoryStats(self: *Self) !void {
        self.mem_info = try memory.MemoryInfo.new();
        self.status = blk: {
            if (self.mem_info.available_ram_percent <= config.free_ram_threshold) {
                const psi = try pressure.pressureSomeAvg10(self.buffer);
                std.log.warn("read avg10: {}", .{psi});
                break :blk MemoryStatus{ .near_terminal = psi };
            } else {
                break :blk MemoryStatus.ok;
            }
        };
    }

    fn freeUpMemory(self: *Self) !void {
        const victim_process = try process.findVictimProcess(self.buffer);

        // Check for memory stats again to see if the
        // low-memory situation was solved while
        // we were searching for our victim
        try self.updateMemoryStats();
        if (self.isMemoryLow()) {
            try victim_process.terminateSelf();
        }
    }

    pub fn poll(self: *Self) !void {
        while (true) {
            if (self.isMemoryLow()) {
                try self.freeUpMemory();
            }

            try self.updateMemoryStats();
            const sleep_time = self.sleepTimeNs();
            std.log.warn("sleeping for {}ms, {}% of RAM is free", .{ sleep_time, self.mem_info.available_ram_percent });

            // Convert ms to ns
            time.sleep(sleep_time * 1000000);
        }
    }

    /// Determines for how long buztd should sleep
    /// This function is essentially a copy of how earlyoom calculates its sleep time
    ///
    /// Credits: https://github.com/rfjakob/earlyoom/blob/dea92ae67997fcb1a0664489c13d49d09d472d40/main.c#L365
    /// MIT Licensed
    fn sleepTimeNs(self: *const Self) u64 {
        // Maximum expected memory fill rate as seen
        // with `stress -m 4 --vm-bytes 4G`
        const ram_fill_rate: i64 = 6000;
        // Maximum expected swap fill rate as seen
        // with membomb on zRAM
        const swap_fill_rate: i64 = 800;

        // Maximum and minimum sleep times (in ms)
        const min_sleep: i64 = 100;
        const max_sleep: i64 = 1000;

        // TODO: make these percentages configurable by args./config. file
        const ram_terminal_percent: f64 = 10.0;
        const swap_terminal_percent: f64 = 10.0;

        const f_ram_headroom_kib = (@intToFloat(f64, self.mem_info.available_ram_percent) - ram_terminal_percent) * (@intToFloat(f64, self.mem_info.total_ram_mb) * 10.0);
        const f_swap_headroom_kib = (@intToFloat(f64, self.mem_info.available_swap_percent) - swap_terminal_percent) * (@intToFloat(f64, self.mem_info.total_swap_mb) * 10.0);

        const i_ram_headroom_kib = math.max(0, @floatToInt(i64, f_ram_headroom_kib));
        const i_swap_headroom_kib = math.max(0, @floatToInt(i64, f_swap_headroom_kib));

        var time_to_sleep = @divFloor(i_ram_headroom_kib, ram_fill_rate) + @divFloor(i_swap_headroom_kib, swap_fill_rate);
        time_to_sleep = math.min(time_to_sleep, max_sleep);
        time_to_sleep = math.max(time_to_sleep, min_sleep);

        return @intCast(u64, time_to_sleep);
    }

    fn isMemoryLow(self: *const Self) bool {
        return switch (self.status) {
            MemoryStatusTag.ok => false,
            MemoryStatusTag.near_terminal => |psi| psi >= config.cutoff_psi,
        };
    }
};
